File: frontend/lib/tapah/request.dart
// ignore_for_file: non_constant_identifier_names

import 'dart:typed_data';

import 'package:frontend/tapah/const.dart';
import 'package:frontend/tapah/class.dart';
import 'package:frontend/tapah/data.dart';
import 'package:frontend/tapah/function.dart';

List<Article> mergeArticlesFromCell(List<Article> existing, String cellValue) {
	final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;
	final urls = cellValue.split(',').map((e) => e.trim()).where((e) => e.isNotEmpty);
	final merged = <Article>[];
	for (final url in urls) {
		Article? old;
		for (final item in existing) {
			if (item.article == url) {
				old = item;
				break;
			}
		}
		merged.add(old ?? Article(article: url, update: now));
	}
	return merged;
}

Future<void> RequestZoneList() async {
	var response = await dio.get(parseurl(url_query_zonelist));
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
	var json = response.data["data"]["zonelist"];
	zonelist = [];
	json.forEach((key, item) {
		zonelist.add(Zone(id: int.parse(key), value: item));
	});
}

Future<void> RequestSectorList() async {
	var response = await dio.get(parseurl(url_query_sectorlist));
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
	var json = response.data["data"]["sectorlist"];
	sectorlist = [];
	json.forEach((key, item) {
		sectorlist.add(Sector(id: int.parse(key), value: item));
	});
}

Future<void> RequestLevelList() async {
	var response = await dio.get(parseurl(url_query_levellist));
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
	var json = response.data["data"]["levellist"];
	levellist = [];
	json.forEach((key, item) {
		levellist.add(Level(id: int.parse(key), value: item));
	});
}

Future<void> RequestFieldList() async {
	var response = await dio.get(parseurl(url_query_fieldlist));
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
	var json = response.data["data"]["fieldlist"];
	fieldlist = [];
	json.forEach((item) {
		fieldlist.add(Field(
			id: item["id"],
			value: item["name"],
			type: item["type"],
			star: item["star"],
			content: item["content"],
			mapping: List<String>.from((item["mapping"] as List?)?.map((e) => e.toString()) ?? [])
				.map((e) => e.trim())
				.where((e) => e.isNotEmpty)
				.toList(),
		));
	});
}

Future<void> RequestEnterpriseList(int page) async {
	var response = await dio.post(parseurl(url_query_enterprise), data: {
		"zone": 0,
		"sector": 0,
		"level": 0,
		"enttype": 0,
		"financial": null,
		"field": 0,
		"name": "",
		"page": page,
	}, options: options,);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
	var json = response.data["data"]["enterpriselist"];
	if (page == 1) enterpriselist = [];
	json.forEach((item) {
		Enterprise enterprise = Enterprise(id: item["id"]);
		enterprise.zone = zonelist.firstWhere((e) => e.id == item["zone"]);
		enterprise.city = item["city"];
		enterprise.name = item["name"];
		enterprise.shortname = item["shortname"];
		enterprise.englishname = item["englishname"];
		enterprise.brief = item["brief"];
		enterprise.upper = item["upper"];
		enterprise.sector = sectorlist.firstWhere((e) => e.id == item["sector"], orElse: () => Sector(id: 0, value: ""));
		enterprise.level = levellist.firstWhere((e) => e.id == item["level"], orElse: () => Level(id: 0, value: ""));
		item["field"].forEach((field) {
			enterprise.fields.add(fieldlist.firstWhere((e) => e.id == field, orElse: () => Field(id: 0, value: "", type: "", star: 0, content: "", mapping: [])));
		});
		enterprise.tags = item["tag"].split(',');
		enterprise.website1 = item["website1"];
		enterprise.website2 = item["website2"];
		enterprise.icon = item["icon"];
		(item["images"] as String).split(',').forEach((image) {
			if (image.isNotEmpty) enterprise.images.add(image);
		});
		if (item["enttype"] ==  '国有企业') enterprise.enttype = 1;
		if (item["enttype"] == '中央企业') enterprise.enttype = 2;
		if (item["financial"] == '是') enterprise.financial = true;
		if (item["financial"] == '否') enterprise.financial = false;
		for (int i = 0;i < item['article1'].length;i++) {
			var article = item['article1'][i];
			var url = (article is Map ? article["article"] : article[0])?.toString() ?? "";
			var update = article is Map ? article["update"] : article[1];
			if (url.isNotEmpty) enterprise.article1.add(Article(article: url, update: update is int ? update : int.tryParse("$update") ?? 0));
		}
		for (int i = 0;i < item['article2'].length;i++) {
			var article = item['article2'][i];
			var url = (article is Map ? article["article"] : article[0])?.toString() ?? "";
			var update = article is Map ? article["update"] : article[1];
			if (url.isNotEmpty) enterprise.article2.add(Article(article: url, update: update is int ? update : int.tryParse("$update") ?? 0));
		}
		enterprise.mapping = item["mapping"];
		enterpriselist.add(enterprise);
	});
}

Future<void> RequestAddZone(String value) async {
	var zone = []; zone.add(value);
	var response = await dio.post(
		parseurl(url_add_zone),
		data: zone,
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestEditZone(Zone zone) async {
	var response = await dio.post(
		parseurl(url_edit_zone), data: {
			"id": zone.id,
			"zone": zone.value,
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestDeleteZone(int id) async {
	var response = await dio.post(parseurl(url_delete_zone), data: {
		"id": id,
	}, options: options,);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestAddSector(String value) async {
	var response = await dio.post(
		parseurl(url_add_sector),
		data: [value],
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestEditSector(Sector sector) async {
	var response = await dio.post(
		parseurl(url_edit_sector), data: {
			"id": sector.id,
			"sector": sector.value,
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestDeleteSector(int id) async {
	var response = await dio.post(
		parseurl(url_delete_sector), data: {
			"id": id,
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestAddLevel(String value) async {
	var response = await dio.post(
		parseurl(url_add_level),
		data: [value],
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestEditLevel(Level level) async {
	var response = await dio.post(
		parseurl(url_edit_level), data: {
			"id": level.id,
			"level": level.value,
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestDeleteLevel(int id) async {
	var response = await dio.post(
		parseurl(url_delete_level), data: {
			"id": id,
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestAddField(String value, String sector, int star, String content) async {
	var response = await dio.post(
		parseurl(url_add_field),
		data: [{
			"name": value,
			"sector": sector,
			"star": star,
			"content": content,
		}],
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestEditField(Field field) async {
	var response = await dio.post(
		parseurl(url_edit_field),
		data: {
			"id": field.id,
			"field": field.value,
			"type": field.type,
			"star": field.star,
			"content": field.content,
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestDeleteField(int id) async {
	var response = await dio.post(
		parseurl(url_delete_field),
		data: {
			"id": id,
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestQuestionListAll() async {
	questionlist = [];
	for (final agent in ['resume', 'joblevel']) {
		var response = await dio.get(parseurl(url_query_questions), queryParameters: {
			"agent": agent,
		});
		if (response.data['code'] != 0) {
			throw Exception('Error code: ${response.data['code']}');
		}
		var json = response.data["data"]["questions"];
		json.forEach((item) {
			questionlist.add(Question(
				id: item["id"],
				agent: item["agent"],
				question: item["question"],
			));
		});
	}
}

Future<void> RequestAddQuestion(String agent, String question) async {
	var response = await dio.post(
		parseurl(url_add_question),
		data: [{
			"agent": agent,
			"question": question,
		}],
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestEditQuestion(Question question) async {
	var response = await dio.post(
		parseurl(url_edit_question),
		data: {
			"id": question.id,
			"agent": question.agent,
			"question": question.question,
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestDeleteQuestion(int id) async {
	var response = await dio.post(
		parseurl(url_delete_question),
		data: {
			"id": id,
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestAddEnterprise(Enterprise enterprise) async {
	var response = await dio.post(
		parseurl(url_add_enterprise),
		data: {
			"name": enterprise.name,
			"shortname": enterprise.shortname,
			"englishname": enterprise.englishname,
			"zone": zonelist.firstWhere((z) => z.id == enterprise.zone?.id).value,
			"city": enterprise.city,
			"brief": enterprise.brief,
			"upper": enterprise.upper,
			"sector": sectorlist.firstWhere((s) => s.id == enterprise.sector?.id).value,
			"level": levellist.firstWhere((l) => l.id == enterprise.level?.id).value,
			"field": enterprise.mapping ?? "",
			"tag": enterprise.tags.join(','),
			"website1": enterprise.website1,
			"website2": enterprise.website2,
			"icon": enterprise.icon,
			"images": enterprise.images.join(','),
			"enttype": enterprise.enttype,
			"financial": enterprise.financial,
			"article1": enterprise.article1.map((e) => [e.article, e.update]).toList(),
			"article2": enterprise.article2.map((e) => [e.article, e.update]).toList(),
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<void> RequestEditEnterprise(Enterprise enterprise) async {
	var response = await dio.post(
		parseurl(url_edit_enterprise),
		data: {
			"id": enterprise.id,
			"name": enterprise.name,
			"shortname": enterprise.shortname,
			"englishname": enterprise.englishname,
			"zone": zonelist.firstWhere((z) => z.id == enterprise.zone?.id).value,
			"city": enterprise.city,
			"brief": enterprise.brief,
			"upper": enterprise.upper,
			"sector": sectorlist.firstWhere((s) => s.id == enterprise.sector?.id).value,
			"level": levellist.firstWhere((l) => l.id == enterprise.level?.id).value,
			"field": enterprise.mapping ?? "",
			"tag": enterprise.tags.join(','),
			"website1": enterprise.website1,
			"website2": enterprise.website2,
			"icon": enterprise.icon,
			"images": enterprise.images.join(','),
			"enttype": enterprise.enttype,
			"financial": enterprise.financial,
			"article1": enterprise.article1.map((e) => e.article).toList(),
			"article2": enterprise.article2.map((e) => e.article).toList(),
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']} status: ${response.data['status']}');
	}
}

Future<void> RequestDeleteEnterprise(int id) async {
	var response = await dio.post(
		parseurl(url_delete_enterprise),
		data: {
			"id": id,
		},
		options: options,
	);
	if (response.data['code'] != 0) {
		throw Exception('Error code: ${response.data['code']}');
	}
}

Future<dynamic> RequestImport(String filename, Uint8List? filedata) async {
	var response = await dio.post(
		parseurl(url_import),
		data: {
			"filename": filename,
			"filedata": String.fromCharCodes(filedata ?? []),
		},
		options: options,
	);
	return response;
}

File: vue/src/tapah/request.ts
import {
	url_query_zonelist,
	url_query_levellist,
	url_query_sectorlist,
	url_query_fieldlist,
	url_query_enterprise,
	url_query_enterprise_detail,
	url_query_case,
	url_query_case_display,
	url_query_article1,
	url_query_article2,
	url_query_wxcode,
	url_query_userinfo,
	url_query_favorite,
	url_query_chatai_auth,
	url_query_chatai_chat,
	url_query_aichat_history,
	url_query_questions,
	miniappid,
} from "./const";
import {
	Zone,
	Sector,
	Level,
	Field,
	Enterprise,
	Case,
	Article,
	AccountInfo,
	ChatItem,
} from "./class";
import {
	zonelist,
	sectorlist,
	levellist,
	fieldlist,
	myfieldlist,
	enterpriselist,
	myenterpriselist,
	caselist,
	article1,
	article2,
	accountinfo,
	chataiToken,
	chataiTokenExpiresAt,
	chataiConversationId,
	chataiAgent,
} from "./data";
import { parseurl } from "./function";

function request<T = any>(options: {
	url: string;
	method?: "GET" | "POST";
	data?: any;
	header?: any;
}): Promise<{ code: number; status?: string; data: T }> {
	return new Promise((resolve, reject) => {
		uni.request({
			url: options.url,
			method: options.method || "GET",
			data: options.data,
			header: options.header,
			success: (res) => {
				if (res.statusCode === 200) {
					resolve(res.data as any);
				} else {
					reject(new Error(`HTTP Error: ${res.statusCode}`));
				}
			},
			fail: (err) => {
				reject(err);
			},
		});
	});
}

function _splitCsv(value: any): string[] {
	if (value === null || value === undefined) return [];
	if (Array.isArray(value)) {
		return value.map((e) => String(e).trim()).filter((e) => e.length > 0);
	}
	return String(value)
		.split(",")
		.map((e) => e.trim())
		.filter((e) => e.length > 0);
}

function parseEnterpriseItem(item: any): Enterprise {
	const enterprise = new Enterprise({ id: item.id });
	enterprise.zone = zonelist.value.find((e) => e.id === item.zone);
	enterprise.city = item.city;
	enterprise.name = item.name;
	enterprise.shortname = item.shortname;
	enterprise.englishname = item.englishname;
	enterprise.brief = item.brief;
	enterprise.upper = item.upper;
	enterprise.sector = sectorlist.value.find((e) => e.id === item.sector);
	enterprise.level = levellist.value.find((e) => e.id === item.level);

	if (Array.isArray(item.field)) {
		item.field.forEach((fId: any) => {
			const f = fieldlist.value.find((e) => e.id === fId);
			if (f) enterprise.fields.push(f);
		});
	}

	enterprise.tags = _splitCsv(item.tag);
	enterprise.website1 = item.website1;
	enterprise.website2 = item.website2;
	enterprise.icon = item.icon;
	enterprise.images = _splitCsv(item.images);

	if (item.enttype !== "国企") enterprise.enttype = 1;
	if (item.enttype === "央企") enterprise.enttype = 2;
	enterprise.financial = item.financial === "是";
	enterprise.growth = item.growth === "是";

	if (Array.isArray(item.article1)) {
		enterprise.article1 = item.article1.map((article: any) => Article.fromJson(article));
	}
	if (Array.isArray(item.article2)) {
		enterprise.article2 = item.article2.map((article: any) => Article.fromJson(article));
	}
	return enterprise;
}

export async function RequestZoneList(): Promise<void> {
	const response = await request<{ zonelist: Record<string, string> }>({
		url: parseurl(url_query_zonelist),
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code}`);
	}
	const json = response.data.zonelist;
	const list = [new Zone({ id: 0, value: "不限" })];
	Object.entries(json).forEach(([key, item]) => {
		list.push(new Zone({ id: parseInt(key, 10), value: item }));
	});
	zonelist.value = list;
}

export async function RequestSectorList(): Promise<void> {
	const response = await request<{ sectorlist: Record<string, string> }>({
		url: parseurl(url_query_sectorlist),
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const json = response.data.sectorlist;
	const list = [new Sector({ id: 0, value: "不限" })];
	Object.entries(json).forEach(([key, item]) => {
		list.push(new Sector({ id: parseInt(key, 10), value: item }));
	});
	sectorlist.value = list;
}

export async function RequestLevelList(): Promise<void> {
	const response = await request<{ levellist: Record<string, string> }>({
		url: parseurl(url_query_levellist),
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const json = response.data.levellist;
	const list = [new Level({ id: 0, value: "不限" })];
	Object.entries(json).forEach(([key, item]) => {
		list.push(new Level({ id: parseInt(key, 10), value: item }));
	});
	levellist.value = list;
}

export async function RequestFieldList(): Promise<void> {
	const response = await request<{ fieldlist: any[] }>({
		url: parseurl(url_query_fieldlist),
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const json = response.data.fieldlist;
	const list: Field[] = [];
	json.forEach((item) => {
		list.push(
			new Field({
				id: item.id,
				value: item.name,
				mapping: Array.isArray(item.mapping) ? item.mapping.map(String) : [],
				type: item.type,
				star: item.star,
				content: item.content,
			})
		);
	});
	fieldlist.value = list;
}

export type FilterParam = number | number[];

export function serializeFilterParam(ids: number[]): FilterParam {
	if (ids.length === 0) return 0;
	if (ids.length === 1) return ids[0];
	return ids;
}

export async function RequestEnterpriseList(
	zone: FilterParam,
	sector: FilterParam,
	level: FilterParam,
	enttype: number,
	field: number,
	financial: boolean | null,
	growth: boolean | null,
	name: string,
	page: number
): Promise<[number, number]> {
	const response = await request<{ enterpriselist: any[]; pagesize: any }>({
		url: parseurl(url_query_enterprise),
		method: "POST",
		data: { zone, sector, level, enttype, field, financial, growth, name, page },
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const data = response.data;
	const json = data.enterpriselist;
	const pagesize = typeof data.pagesize === "number" ? data.pagesize : parseInt(String(data.pagesize), 10) || 20;

	const list = [...enterpriselist.value];
	json.forEach((item) => {
		list.push(parseEnterpriseItem(item));
	});
	enterpriselist.value = list;

	return [json.length, pagesize];
}

export async function RequestEnterprise(
	zone: FilterParam,
	sector: FilterParam,
	level: FilterParam,
	enttype: number,
	field: number,
	financial: boolean | null,
	growth: boolean | null,
	name: string,
	page: number
): Promise<Enterprise[]> {
	const response = await request<{ enterpriselist: any[] }>({
		url: parseurl(url_query_enterprise),
		method: "POST",
		data: { zone, sector, level, enttype, field, financial, growth, name, page },
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const json = response.data.enterpriselist;
	return json.map(parseEnterpriseItem);
}

export async function RequestEnterpriseDetail(id: number): Promise<Enterprise> {
	const response = await request<{ enterprise: any }>({
		url: parseurl(url_query_enterprise_detail),
		method: "POST",
		data: { id },
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const item = response.data.enterprise;
	return parseEnterpriseItem(item);
}

export async function RequestArticle1(): Promise<number> {
	const response = await request<{ link: any[] }>({
		url: parseurl(url_query_article1),
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const json = response.data.link;
	const list = json.map((item) => Article.fromJson(item));
	list.sort((a, b) => b.update - a.update);
	article1.value = list;
	return json.length;
}

export async function RequestArticle2(): Promise<number> {
	const response = await request<{ link: any[] }>({
		url: parseurl(url_query_article2),
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const json = response.data.link;
	article2.value = json.map((item) => Article.fromJson(item));
	return json.length;
}

export async function RequestCaseList(
	enterprise: number,
	level: number,
	sector: number,
	field: number,
	stag1: number,
	stag2: number,
	year: number,
	page: number
): Promise<[number, number]> {
	const response = await request<{ caselist: any[]; pagesize: any }>({
		url: parseurl(url_query_case),
		method: "POST",
		data: { enterprise, level, sector, field, stag1, stag2, year, page },
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const data = response.data;
	const json = data.caselist;
	const pagesize = typeof data.pagesize === "number" ? data.pagesize : parseInt(String(data.pagesize), 10) || 20;

	const list = [...caselist.value];
	json.forEach((item) => {
		const c = new Case({ id: item.id, name: item.name });
		c.entid = item.entid;
		c.enticon = item.enticon;
		c.entname = item.entname;
		c.field = fieldlist.value.find((e) => e.id === item.field);
		c.tags = Array.isArray(item.tags) ? item.tags.map(String) : [];
		c.student = item.student;
		c.school1 = item.school1;
		c.stag1 = item.stag1;
		c.field1 = item.field1;
		c.school2 = item.school2;
		c.stag2 = item.stag2;
		c.field2 = item.field2;
		c.year = item.year;
		c.detail = item.detail;
		c.dep = item.dep;
		list.push(c);
	});
	caselist.value = list;
	return [json.length, pagesize];
}

function parseCaseItem(item: any): Case {
	const c = new Case({ id: item.id, name: item.name });
	c.entid = item.entid;
	c.enticon = item.enticon;
	c.entname = item.entname;
	c.field = fieldlist.value.find((e) => e.id === item.field);
	c.tags = Array.isArray(item.tags) ? item.tags.map(String) : [];
	c.student = item.student;
	c.school1 = item.school1;
	c.stag1 = item.stag1;
	c.field1 = item.field1;
	c.school2 = item.school2;
	c.stag2 = item.stag2;
	c.field2 = item.field2;
	c.year = item.year;
	c.detail = item.detail;
	c.dep = item.dep;
	return c;
}

export interface CaseDisplayResult {
	primary: Case | null;
	similar: Case[];
	similarTotal: number;
	pageSize: number;
}

export async function RequestCaseDisplay(
	enterprise: number,
	level: number,
	sector: number,
	field: number,
	stag1: number,
	stag2: number,
	year: number,
	page: number,
	options?: { excludeId?: number; referenceId?: number },
): Promise<CaseDisplayResult> {
	const response = await request<{ primary: any; similarlist: any[]; similar_total: number; pagesize: number }>({
		url: parseurl(url_query_case_display),
		method: "POST",
		data: {
			enterprise,
			level,
			sector,
			field,
			stag1,
			stag2,
			year,
			page,
			exclude_id: options?.excludeId ?? 0,
			reference_id: options?.referenceId ?? 0,
		},
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const data = response.data;
	const primary = data.primary ? parseCaseItem(data.primary) : null;
	const similar = (data.similarlist ?? []).map(parseCaseItem);
	return {
		primary,
		similar,
		similarTotal: data.similar_total ?? 0,
		pageSize: typeof data.pagesize === "number" ? data.pagesize : parseInt(String(data.pagesize), 10) || 20,
	};
}

export async function RequestCase(
	enterprise: number,
	level: number,
	sector: number,
	field: number,
	stag1: number,
	stag2: number,
	year: number,
	page: number
): Promise<Case[]> {
	const response = await request<{ caselist: any[] }>({
		url: parseurl(url_query_case),
		method: "POST",
		data: { enterprise, level, sector, field, stag1, stag2, year, page },
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const json = response.data.caselist;
	return json.map(parseCaseItem);
}

export async function RequestWxCode(code: string): Promise<void> {
	const response = await request<any>({
		url: parseurl(url_query_wxcode),
		method: "POST",
		data: { code },
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}
	const json = response.data;
	const info = new AccountInfo();
	info.id = json.id;
	info.openid = json.openid;
	info.nickname = json.nickname || "微信名称";
	if (info.nickname.length === 0) {
		info.nickname = "微信名称";
	}
	info.avatar = json.avatar || "";
	info.field = new Set(Array.isArray(json.field) ? json.field.map(Number) : []);
	info.enterprise = new Set(Array.isArray(json.enterprise) ? json.enterprise.map(Number) : []);
	accountinfo.value = info;
}

export async function RequestUserInfo(): Promise<void> {
	if (!accountinfo.value) {
		throw new Error("User not logged in");
	}
	await request<void>({
		url: parseurl(url_query_userinfo),
		method: "POST",
		data: {
			openid: accountinfo.value.openid,
			avatar: accountinfo.value.avatar,
			nickname: accountinfo.value.nickname,
			field: Array.from(accountinfo.value.field),
			enterprise: Array.from(accountinfo.value.enterprise),
		},
	});
}

export async function RequestFavorite(): Promise<void> {
	if (!accountinfo.value) {
		throw new Error("User not logged in");
	}
	const response = await request<{ field: any[]; enterprise: any[] }>({
		url: parseurl(url_query_favorite),
		method: "POST",
		data: {
			enterprise: Array.from(accountinfo.value.enterprise),
			field: Array.from(accountinfo.value.field),
		},
	});
	if (response.code !== 0) {
		throw new Error(`Error code: ${response.code} status: ${response.status}`);
	}

	const json1 = response.data.field;
	const list1: Field[] = [];
	json1.forEach((item) => {
		list1.push(
			new Field({
				id: item.id,
				value: item.name,
				mapping: Array.isArray(item.mapping) ? item.mapping.map(String) : [],
				type: item.type,
				star: item.star,
				content: item.content,
			})
		);
	});
	myfieldlist.value = list1;

	const json2 = response.data.enterprise;
	const list2: Enterprise[] = json2.map(parseEnterpriseItem);
	myenterpriselist.value = list2;
}

export async function RequestChatAIAuth(): Promise<void> {
	if (!accountinfo.value) {
		throw new Error("User not logged in");
	}
	const response = await request<any>({
		url: parseurl(url_query_chatai_auth),
		method: "POST",
		data: {
			openid: accountinfo.value.openid,
			unionid: accountinfo.value.unionid,
			nickname: accountinfo.value.nickname,
			avatar: accountinfo.value.avatar,
			phone: accountinfo.value.openid,
			email: accountinfo.value.email,
			realname: accountinfo.value.realname,
			useridentity: accountinfo.value.useridentity,
			usersource: accountinfo.value.usersource,
			miniappid: miniappid,
		},
	});
	if (response.code !== 0) {
		throw new Error(response.status || "鉴权失败");
	}
	const json = response.data;
	chataiToken.value = json.token;
	chataiTokenExpiresAt.value = json.expiresAt;
}

export async function RequestChatAIChat(message: string, agent?: string): Promise<string> {
	if (!accountinfo.value) {
		throw new Error("User not logged in");
	}
	if (!chataiToken.value) {
		throw new Error("未鉴权");
	}
	const agentKey = agent || chataiAgent.value;
	const data: Record<string, any> = {
		openid: accountinfo.value.openid,
		token: chataiToken.value,
		message: message,
		agent: agentKey,
	};
	if (chataiConversationId.value) {
		data.conversationId = chataiConversationId.value;
	}
	const response = await request<any>({
		url: parseurl(url_query_chatai_chat),
		method: "POST",
		data: data,
	});
	if (response.code !== 0) {
		throw new Error(response.status || "对话失败");
	}
	const json = response.data;
	chataiConversationId.value = json.conversationId;
	chataiAgent.value = agentKey;
	return json.response || "";
}

export class AIChatHistoryResult {
	messages: ChatItem[];
	hasMore: boolean;
	constructor({ messages, hasMore }: { messages: ChatItem[]; hasMore: boolean }) {
		this.messages = messages;
		this.hasMore = hasMore;
	}
}

export async function RequestAIChatHistory({
	agent,
	before,
	limit = 10,
}: {
	agent?: string;
	before?: number;
	limit?: number;
}): Promise<AIChatHistoryResult> {
	if (!accountinfo.value) {
		throw new Error("User not logged in");
	}
	const agentKey = agent || chataiAgent.value;
	const queryParameters: Record<string, any> = {
		openid: accountinfo.value.openid,
		agent: agentKey,
		limit: limit,
	};
	if (before !== undefined) {
		queryParameters.before = before;
	}

	const queryString = Object.entries(queryParameters)
		.map(([key, val]) => `${encodeURIComponent(key)}=${encodeURIComponent(String(val))}`)
		.join("&");
	const url = `${parseurl(url_query_aichat_history)}?${queryString}`;

	const response = await request<any>({ url });
	if (response.code !== 0) {
		throw new Error(response.status || "加载聊天记录失败");
	}
	const json = response.data;
	if (before === undefined) {
		chataiConversationId.value = json.conversationId;
		chataiAgent.value = agentKey;
	}
	const list: ChatItem[] = [];
	for (const item of json.messages || []) {
		list.push(
			new ChatItem({
				isuser: item.isuser === true,
				detail: item.detail || "",
				timestamp: item.timestamp || 0,
			})
		);
	}
	return new AIChatHistoryResult({
		messages: list,
		hasMore: json.hasMore === true,
	});
}

export async function RequestQuestions(agent: string): Promise<string[]> {
	const url = `${parseurl(url_query_questions)}?agent=${encodeURIComponent(agent)}`;
	const response = await request<any>({ url });
	if (response.code !== 0) {
		throw new Error(response.status || "加载问题失败");
	}
	const list: string[] = [];
	for (const item of response.data.questions || []) {
		const question = item.question?.toString() || "";
		if (question) list.push(question);
	}
	return list;
}
